### PR TITLE
mgr/stats: initialize mx_last_updated in FSPerfStats

### PR DESCRIPTION
**Pull request by joscollin**
Fixes: https://tracker.ceph.com/issues/65073

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by joscollin** [comment]
jenkins test make check arm64

**Comment by vshankar** [comment]
@joscollin I'm not understanding how a bug in mgr/stats is fixed by patching test_fstop.py. By reading the commit message it looks to me that this is an optimization, so, I'm assuming a real fix will be followed up shortly?

**Comment by joscollin** [comment]
> shortly

@vshankar Fixed.

File: src/pybind/mgr/stats/fs/perf_stats.py [hunk 1]
```diff
@@ -201,7 +202,7 @@ def notify_fsmap(self):
         #Reregister the user queries when there is a new rank0 mds
         with self.lock:
             gid_state = FSPerfStats.get_rank0_mds_gid_state(self.module.get('fs_map'))
-            if not gid_state:
+            if not gid_state or not self.mx_last_updated:
```
Comment (neesingh-rh) [inline src/pybind/mgr/stats/fs/perf_stats.py:205]: Agreed! it will prevent unnecessary notify calls.

**Review by neesingh-rh — APPROVED** [review]

File: src/pybind/mgr/stats/fs/perf_stats.py (hunk 1 above)
Comment (vshankar) [inline src/pybind/mgr/stats/fs/perf_stats.py:205]: Does this really help? Could you explain why?

AFAICT, initializing `self.mx_last_updated = 0.0` in `FSPerfStats.__init__` should suffice. @joscollin @neesingh-rh

File: qa/tasks/cephfs/test_fstop.py [hunk 2]
```diff
@@ -81,6 +81,9 @@ def test_dumpfs(self):
         """
         Tests 'cephfs-top --dumpfs' output is valid
         """
+        # disable stats plugin for unnecessary notifys
+        self._disable_mgr_stats_plugin()
```
Comment (vshankar) [inline qa/tasks/cephfs/test_fstop.py:85]: I don't have anything against this optimization, but is this _really_ required?

File: src/pybind/mgr/stats/fs/perf_stats.py (hunk 1 above)
Comment (joscollin) [reply src/pybind/mgr/stats/fs/perf_stats.py:205]: There's no need to take it further, no need to start the timer too. So catch it and return early.

File: qa/tasks/cephfs/test_fstop.py (hunk 2 above)
Comment (joscollin) [reply qa/tasks/cephfs/test_fstop.py]: Fixed

File: src/pybind/mgr/stats/fs/perf_stats.py (hunk 1 above)
Comment (vshankar) [reply src/pybind/mgr/stats/fs/perf_stats.py:205]: > There's no need to take it further, no need to start the timer too. So catch it and return early.

That's explaining what the change is and not why the change is required. What can go wrong if the timer kicks in ? In `re_register_queries`:

```
if self.mx_last_updated >= ua_last_updated:
    ....
```

The else path would be taken which would register the timer again. Are you worried about registering the timer over and over again when `self.mx_last_updated == 0`?

Comment (vshankar) [reply src/pybind/mgr/stats/fs/perf_stats.py]: So, the additional check wasn't really required? What about my query in the above comment @joscollin @neesingh-rh

Comment (joscollin) [reply src/pybind/mgr/stats/fs/perf_stats.py]: As you mentioned [here](https://github.com/ceph/ceph/pull/56525#discussion_r1546065807), the initialisation is sufficient for resolving this issue. The additional check is not required. The timer usually starts whenever there's an fsmap change, that's normal.

**Comment by joscollin** [comment]
jenkins test make check arm64

**Review by vshankar — APPROVED** [review]

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#2024-05-13

**Pull request merged**